src/org/opensolaris/opengrok/util/AcceptHelper.java

```diff
@@ -0,0 +1,236 @@
+            File t1 = f1;
+            while ((t1 = t1.getParentFile()) != null) {
+                if (f2.equals(t1)) {
+                    LOGGER.log(Level.INFO, "Skipping links to parent...: {0} {1}",
+                            new Object[]{parent.getAbsolutePath(), file.getAbsolutePath()});
+                    return false;
+                }
+            }
+
+            return accept(project, file);
+        } catch (IOException ex) {
+            LOGGER.log(Level.WARNING, "Failed to resolve name: {0} {1}",
+                    new Object[]{parent.getAbsolutePath(), file.getAbsolutePath()});
+        }
+        return false;
+    }
+
+    /**
+     * Check if I should accept the path containing a symlink. This method has
+     * the same effect as the call acceptSymlink(null, file).
+     *
+     * @see AcceptHelper#acceptSymlink(Project, File)
+     * @param file the file symlink
+     * @return true if the file should be accepted, false otherwise
+     * @throws java.io.IOException
+     */
+    public static boolean acceptSymlink(File file) throws IOException {
+        return acceptSymlink(null, file);
+    }
+
+    /**
+     * Check if I should accept the path containing a symlink.
+     *
+     * @param project if the file is relevant to some project (may be null)
+     * @param file the file symlink
+     * @return true if the file should be accepted, false otherwise
+     * @throws java.io.IOException
+     */
+    public static boolean acceptSymlink(Project project, File file) throws IOException {
+        String absolutePath = file.getAbsolutePath(), canonicalPath = file.getCanonicalPath();
+        RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+        // Always accept local symlinks
+        if (isLocal(project, canonicalPath)) {
+            return true;
+        }
+
+        for (String allowedSymlink : env.getAllowedSymlinks()) {
+            if (absolutePath.startsWith(allowedSymlink)) {
+                String allowedTarget = new File(allowedSymlink).getCanonicalPath();
+                if (canonicalPath.startsWith(allowedTarget)
+                        && absolutePath.substring(allowedSymlink.length()).equals(canonicalPath.substring(allowedTarget.length()))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a file is local to the current project. If we don't have
+     * projects, check if the file is in the source root.
+     *
+     * @param project a project (may be null)
+     * @param path the path to a file
+     * @return true if the file is local to the current repository
+     */
+    public static boolean isLocal(Project project, String path) {
+        RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+        String srcRoot = env.getSourceRootPath();
+
+        if (path.startsWith(srcRoot)) {
+            if (env.hasProjects()) {
```

Comment:
Not really here, because then if the `project == null` then the return value is `true` which is not the original intention. I moved the code around to avoid the substring operation.